samples/snippets/sample_custom_types.py

```diff
@@ -0,0 +1,527 @@
+#  Copyright 2021 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+# [START compute_custom_machine_type_create ]
+from collections import namedtuple
+from enum import Enum, unique
+import sys
+import time
+from typing import Union
+
+from google.cloud import compute_v1
+
+
+# [END compute_custom_machine_type_create ]
+
+
+# [START compute_custom_machine_type_helper_class ]
+def gb_to_mb(value: int) -> int:
+    return value << 10
+
+
+class CustomMachineType:
+    """
+    Allows to create custom machine types to be used with the VM instances.
+    """
+
+    @unique
+    class CPUSeries(Enum):
+        N1 = "custom"
+        N2 = "n2-custom"
+        N2D = "n2d-custom"
+        E2 = "e2-custom"
+        E2_MICRO = "e2-custom-micro"
+        E2_SMALL = "e2-custom-small"
+        E2_MEDIUM = "e2-custom-medium"
+
+    TypeLimits = namedtuple(
+        "TypeLimits",
+        [
+            "allowed_cores",
+            "min_mem_per_core",
+            "max_mem_per_core",
+            "allow_extra_memory",
+            "extra_memory_limit",
+        ],
+    )
+
+    LIMITS = {
+        CPUSeries.E2: TypeLimits(range(2, 33, 2), 512, 8192, False, 0),
+        CPUSeries.E2_MICRO: TypeLimits([], 1024, 2048, False, 0),
+        CPUSeries.E2_SMALL: TypeLimits([], 2048, 4096, False, 0),
+        CPUSeries.E2_MEDIUM: TypeLimits([], 4096, 8192, False, 0),
+        CPUSeries.N2: TypeLimits(
+            set(range(2, 33, 2)).union(set(range(36, 129, 4))),
+            512,
+            8192,
+            True,
+            gb_to_mb(624),
+        ),
+        CPUSeries.N2D: TypeLimits(
+            [2, 4, 8, 16, 32, 48, 64, 80, 96], 512, 8192, True, gb_to_mb(768)
+        ),
+        CPUSeries.N1: TypeLimits(
+            {1}.union(range(2, 97, 2)), 922, 6656, True, gb_to_mb(624)
+        ),
+    }
+
+    def __init__(
+        self, zone: str, cpu_series: CPUSeries, memory_mb: int, core_count: int = 0
+    ):
+        self.zone = zone
+        self.cpu_series = cpu_series
+        self.limits = self.LIMITS[self.cpu_series]
+        self.core_count = 2 if self.is_shared() else core_count
+        self.memory_mb = memory_mb
+        self._check()
+        self.extra_memory_used = self._check_extra_memory()
+
+    def is_shared(self):
+        return self.cpu_series in (
+            self.CPUSeries.E2_SMALL,
```

Comment:
These three aren't attributes of self, just CPUSeries.E2_SMALL, etc.

Reply:
OK, it's not attribute of `self`, but `self` was easier to write than the whole name of the class `CustomMachineType`. I'll change it to `CustomMachineType` though, so it's nicer.

Just `CPUSeries.E2_SMALL` wouldn't work in this context: `NameError: name 'CPUSeries' is not defined`